src/emulator/mcardGCN.c

```diff
@@ -73,25 +73,369 @@ static s32 bWrite2Card;
+}
+
+static bool mcardCheckChecksumFileHeader(MemCard* pMCard, char* buffer) {
+    s32 pad2;
+    s32 checksum;
+    char buffer2[8192];
+    s32 toggle = 1;
+    s32 pad[2];
+
+    memcpy(buffer2, pMCard->writeBuffer, 0x2000);
+    memcpy(pMCard->writeBuffer, buffer, 0x2000);
+
+    mcardCalculateChecksumFileBlock1(pMCard, &checksum);
+
+    if (checksum != *(s32*)(pMCard->writeBuffer + 0x28)) {
+        toggle = 0;
+        if (mcardReplaceFileBlock(pMCard, 0) == false) {
+            return false;
+        }
+    }
+    memcpy(pMCard->writeBuffer, buffer + 0x2000, 0x2000);
+
+    mcardCalculateChecksumFileBlock2(pMCard, &checksum);
+
+    if (checksum != *(s32*)(pMCard->writeBuffer + 0x106C)) {
+        if (toggle == 1) {
+            if (mcardReplaceFileBlock(pMCard, 1) == false) {
+                return false;
+            }
+        } else {
+            pMCard->error = MC_E_CHECKSUM;
+            return false;
+        }
+    }
+    memcpy(pMCard->writeBuffer, buffer2, 0x2000);
+
+    return true;
+}
 
 static bool mcardPoll(MemCard* pMCard);
-#pragma GLOBAL_ASM("asm/non_matchings/mcardGCN/mcardPoll.s")
+
+static inline bool UnkInlinemCardVerifyChecksumFileHeader(MemCard* pMCard) {
+    if (pMCard->saveToggle == true) {
+        if (mcardReadyCard(pMCard) == false) {
+            return false;
+        }
+        if (mcardGCErrorHandler(pMCard, CARDOpen(pMCard->slot, pMCard->file.name, &pMCard->file.fileInfo)) != true) {
+            CARDUnmount(pMCard->slot);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static inline bool UnkInlinemCardVerifyChecksumFileHeader2(MemCard* pMCard, char* buffer) {
+    if (mCard.saveToggle == true) {
+        if (mcardGCErrorHandler(pMCard, CARDReadAsync(&pMCard->file.fileInfo, buffer, 0x6000, 0, NULL)) != true) {
+            return false;
+        }
+        if (mcardPoll(pMCard) != true) {
+            return false;
+        }
+        DCInvalidateRange(buffer, 0x6000);
+    }
+
+    return true;
+}
+
+static inline bool UnkINlinemCardVerifyChecksumFileHeader3(MemCard* pMCard, char** buffer) {
+    if (xlHeapFree(buffer) == false) {
+        return false;
+    }
+    if (pMCard->saveToggle == true) {
+        if (pMCard->file.fileInfo.chan != -1) {
+            CARDClose(&pMCard->file.fileInfo);
+        }
+        CARDUnmount(pMCard->slot);
+    }
+    return false;
+}
+
+static inline bool UnkINlinemCardVerifyChecksumFileHeader4(MemCard* pMCard, char** buffer) {
+    if (xlHeapFree(buffer) == false) {
+        return false;
+    }
+    if (pMCard->saveToggle == true) {
+        if (pMCard->file.fileInfo.chan != -1) {
+            CARDClose(&pMCard->file.fileInfo);
+        }
+        CARDUnmount(pMCard->slot);
+    }
+    return true;
+}
+
+static bool mcardVerifyChecksumFileHeader(MemCard* pMCard) {
+    char* buffer;
+
+    if (UnkInlinemCardVerifyChecksumFileHeader(pMCard) == false) {
+        return false;
+    }
+    if (xlHeapTake(&buffer, 0x30006000) == 0) {
```

Comment:
```suggestion
    if (!xlHeapTake(&buffer, 0x6000 | 0x30000000)) {
```
hopefully the ``== 0`` change doesn't break the match (also do that for other functions like this too)